core-bundle/src/Resources/contao/classes/FrontendTemplate.php

```diff
@@ -356,24 +357,48 @@ private function setCacheHeaders(Response $response)
 			return $response->setPrivate();
 		}
 
-		// Do not cache the response if a user is logged in or the page is protected
-		// TODO: Add support for proxies so they can vary on member context
-		if (FE_USER_LOGGED_IN === true || BE_USER_LOGGED_IN === true || $objPage->protected || $this->hasAuthenticatedBackendUser())
+		// Do not cache response if any cookies were received or sent and make sure the response is private.
+		if (0 !== \count($response->headers->getCookies()) || 0 !== \count(System::getContainer()->get('request_stack')->getCurrentRequest()->headers->getCookies()))
```

Comment:
as usualy, very unsure if `getCurrentRequest` is correct, as the cookies are only on the master request and (probably) not added to subrequests.

Reply:
As usually, it is correct :) The master request is very, very rarely the right one. Cookies can also be set on a subrequest and we have to have the very same handling there. Imagine you create an ESI fragment that uses `Template->getResponse()`. It should work just the same.

Reply:
I'm not sure I understand this correctly.
 1. if Symfony's `HttpCache` is used, the *current* request might be a subrequest for ESI, which means cookies are likely not there (because its not a clone of the master request). So using the master request would be better.
 2. if this is a Varnish or other reverse proxy, it's always the master request, so using the master request would be fine?